### PR TITLE
Support Simulcast Stream (RID Based Simulcast)

### PR DESCRIPTION
**Pull request by johzzy**
## Publisher: 

open Simulcast推流 (`/players/rtc_publisher_simulcast.html` page)

## Player: 

open RTC播放器 (`/players/rtc_player.html` page)

    low: webrtc://127.0.0.1/live/livestream?layer=low
    mid: webrtc://127.0.0.1/live/livestream?layer=mid
    high: webrtc://127.0.0.1/live/livestream?layer=high

Related Link: https://github.com/ossrs/srs/pull/2420

### CODE REVIEW

File: trunk/src/app/srs_app_rtc_conn.cpp [hunk 1]
```diff
@@ -1748,6 +1748,14 @@ void SrsRtcPublishStream::update_send_report_time(uint32_t ssrc, const SrsNtp& n
     }
 }
 
+void SrsRtcPublishStream::bind_rid(const SrsRidInfo &rid_info) {
+    // NOTE: get unused video_track, and active as rid info.
+    SrsRtcVideoRecvTrack* video_track = get_video_track(0);
```
Comment (xiaozhihong) [inline trunk/src/app/srs_app_rtc_conn.cpp:1760]: What does '0' mean?

`TRANS_BY_GPT3`

File: trunk/src/app/srs_app_rtc_conn.cpp [hunk 2]
```diff
@@ -3156,6 +3194,9 @@ srs_error_t SrsRtcConnection::generate_publish_local_sdp(SrsRequest* req, SrsSdp
         }
 
         SrsVideoPayload* payload = (SrsVideoPayload*)video_track->media_;
+        if (srs_string_contains(req->param,"?numberOfSimulcastLayers=spec3")) {
+            payload->h264_param_.x_google_bitrate = "x-google-min-bitrate=5000;x-google-max-bitrate=8000;x-google-start-bitrate=6000";
```
Comment (xiaozhihong) [inline trunk/src/app/srs_app_rtc_conn.cpp]: Would it be better for the user to control this? For example, the simulcast.html page can modify the SDP answer itself.

`TRANS_BY_GPT3`

File: trunk/src/app/srs_app_rtc_conn.cpp (hunk 1 above)
Comment (johzzy) [reply trunk/src/app/srs_app_rtc_conn.cpp:1760]: get unused video_track,

Before executing here, multiple tracks will be allocated based on simulcast SDP, with the default SSRC as 0, indicating unused.

When reaching this point, take an unused track and bind it with RID and SSRC.

`TRANS_BY_GPT3`

File: trunk/src/app/srs_app_rtc_sdp.cpp [hunk 3]
```diff
@@ -1177,3 +1209,215 @@ srs_error_t SrsSdp::update_msid(string id)
...
+    // a=extmap:5 urn:ietf:params:rtp-hdrext:sdes:rtp-stream-id
+    // a=extmap:6 urn:ietf:params:rtp-hdrext:sdes:repaired-rtp-stream-id
+
+    for (size_t i=0; i<kExtMapFieldSize; ++i) {
+        if (ext == kExtMapFieldArray[i]) {
+            ext_id_list_[i] = id;
+        }
+    }
+}
+
```
Comment (xiaozhihong) [inline trunk/src/app/srs_app_rtc_sdp.cpp]: Please process the following code according to the SRS method. SRS has a structure called Rtp, which does not use the C memory structure but follows the decode method.

`TRANS_BY_GPT3`

**Review by xiaozhihong — APPROVED** [review]

**Pull request merged**